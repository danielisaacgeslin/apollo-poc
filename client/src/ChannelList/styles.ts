import { css } from 'emotion';

export const item = css`
  margin-bottom: 5px;
`;
