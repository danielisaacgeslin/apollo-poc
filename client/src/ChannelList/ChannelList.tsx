import React, { memo } from 'react';
import { ApolloError } from 'apollo-boost';

import { Channel } from '../../../server/src/channel/ChannelAPI'; /** @todo move to a shared location */
import { DangerButton } from '../shared/DangerButton';
import { item } from './styles';

export interface Props {
  isLoading: boolean;
  isDeleting: boolean;
  error: ApolloError | undefined;
  channelList: Channel[];
  onDelete: (id: string) => void;
}

const ChannelList = ({ channelList = [], onDelete, isLoading, isDeleting, error }: Props) => {
  const onDeleteChannel = (id: string, name: string) => {
    if (window.confirm(`Are you sure you want to delete "${name}"`)) onDelete(id);
  };

  return (
    <div>
      <h4>Channel list</h4>
      <div>
        {isLoading && <p>loading list...</p>}
        {error && <p>ups: {error?.message}</p>}
        {!channelList.length && !isLoading && <p>no results.</p>}
        <div>
          {channelList.map(({ id, name, type }) => (
            <p key={id} className={item}>
              <DangerButton data-testid="channel-delete" disabled={isDeleting} onClick={() => onDeleteChannel(id, name)}>
                x
              </DangerButton>{' '}
              {name} <small>({type})</small>
            </p>
          ))}
        </div>
      </div>
    </div>
  );
};

export default memo(ChannelList);
