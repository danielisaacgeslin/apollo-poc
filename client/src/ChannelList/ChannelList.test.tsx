import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { default as ChannelList, Props } from './ChannelList';

describe('ChannelList', () => {
  let props: Props;

  beforeEach(() => {
    props = {
      isLoading: false,
      isDeleting: false,
      error: undefined,
      channelList: [
        { id: 'a', name: 'first', type: 'dm' },
        { id: 'b', name: 'second', type: 'group' }
      ],
      onDelete: jest.fn()
    };
  });

  it('should render', () => {
    const { container } = render(<ChannelList {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render with no results', () => {
    const { container } = render(<ChannelList {...props} channelList={[]} />);
    expect(container).toMatchSnapshot();
  });

  it('should render loading', () => {
    const { container } = render(<ChannelList {...props} isLoading={true} channelList={[]} />);
    expect(container).toMatchSnapshot();
  });

  it('should render with error', () => {
    const { container } = render(<ChannelList {...props} error={{ message: 'error message' } as any} channelList={[]} />);
    expect(container).toMatchSnapshot();
  });

  it('should NOT delete when NOT confirmed', () => {
    window.confirm = () => false;
    const { getAllByTestId } = render(<ChannelList {...props} />);
    fireEvent.click(getAllByTestId('channel-delete')[0]);
    expect(props.onDelete).not.toBeCalled();
  });

  it('should delete when confirmed', () => {
    window.confirm = () => true;
    const { getAllByTestId } = render(<ChannelList {...props} />);
    fireEvent.click(getAllByTestId('channel-delete')[0]);
    expect(props.onDelete).toBeCalledWith(props.channelList[0].id);
  });

  it('should NOT delete when is deleting', () => {
    window.confirm = () => true;
    const { getAllByTestId } = render(<ChannelList {...props} isDeleting={true} />);
    fireEvent.click(getAllByTestId('channel-delete')[0]);
    expect(props.onDelete).not.toBeCalledWith(props.channelList[0].id);
  });
});
