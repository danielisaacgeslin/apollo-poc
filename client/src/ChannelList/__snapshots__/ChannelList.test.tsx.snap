// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelList should render 1`] = `
.emotion-1 {
  margin-bottom: 5px;
}

.emotion-0 {
  font-size: 13px;
  padding: 3px 10px;
  background-color: #D6DDF6;
  color: #252A2E;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  background-color: #B73F65;
  color: #FFFFFF;
}

<div>
  <div>
    <h4>
      Channel list
    </h4>
    <div>
      <div>
        <p
          class="emotion-1"
        >
          <button
            class="emotion-0"
            data-testid="channel-delete"
          >
            x
          </button>
           
          first
           
          <small>
            (
            dm
            )
          </small>
        </p>
        <p
          class="emotion-1"
        >
          <button
            class="emotion-0"
            data-testid="channel-delete"
          >
            x
          </button>
           
          second
           
          <small>
            (
            group
            )
          </small>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ChannelList should render loading 1`] = `
<div>
  <div>
    <h4>
      Channel list
    </h4>
    <div>
      <p>
        loading list...
      </p>
      <div />
    </div>
  </div>
</div>
`;

exports[`ChannelList should render with error 1`] = `
<div>
  <div>
    <h4>
      Channel list
    </h4>
    <div>
      <p>
        ups: 
        error message
      </p>
      <p>
        no results.
      </p>
      <div />
    </div>
  </div>
</div>
`;

exports[`ChannelList should render with no results 1`] = `
<div>
  <div>
    <h4>
      Channel list
    </h4>
    <div>
      <p>
        no results.
      </p>
      <div />
    </div>
  </div>
</div>
`;
